Brush painting from a image

Zaspard opened the pull request:
<h2>About pull request</h2>

Added possibility to create a brush by image. 

If a brush was created on the image, then:
1)The area is drawn by the image.
2)The area is clipped to the object
3)The object is drawn (if a pen has been set).  

Example:
[HelloImageWorld-B4FEA3FE6CD74639B3180A99F226EDD7_tempfile.pdf](https://github.com/empira/PDFsharp/files/12745452/HelloImageWorld-B4FEA3FE6CD74639B3180A99F226EDD7_tempfile.pdf)

![2023-09-28_11h13_02](https://github.com/empira/PDFsharp/assets/45457436/75a0154f-2d43-4e4f-98a0-8ae3557daa88)


<h2>Changelog</h2>

Added class XImageBrush.cs
For class XImageBrush are realized:

- CORE
- WPF (not tested)
- GDI (not tested)

Added implementation for the XGraphicsPdfRenderer for:
- DrawEllipse
- DrawRectangle
- DrawRoundedRectangle
- DrawPath
- DrawPolygon
- DrawPie (XGraphicsPath.AddPie not implemented)
- DrawClosedCurve (XGraphicsPath.AddClosedCurve not implemented)

Added a test for the implementation.

Zaspard committed:
Init XImageBrush

Zaspard committed:
Implemented XImageBrush class:
-CORE
-GDI (not tested)
-WFP (not tested)
Added test for XImageBrush

Zaspard commented:
Should I update to the latest version or close it?

ThomasHoevel commented:
The image looks good, but "WPF (not tested)" and "GDI (not tested)" in the changelog made me hesitate and I did not evaluate the PR yet.
Did you compile and test WPF and GDI meanwhile?

Zaspard commented:
I compiled, but didn't test it as I don't know how to test it correctly. I haven't found any tests on WPF and GDI for PDFSharp, only for CORE.
Would it be enough to change the reference from PDFSharp to PDFSharp-WPF in the tests and just run (same for GDI)?

ThomasHoevel commented:
In most cases, the .cs files exist only in the Core folder, but are linked by the GDI and WPF projects.  
The implementations are different, but the tests should be the same (except maybe setting the FontResolver for Core tests only).

Zaspard commented:
Thanks for the reply, I'll try to get it done this week

Zaspard committed:
Merge branch 'master' of https://github.com/empira/PDFsharp into XImageBrush

Conflicts:
	src/foundation/src/PDFsharp/src/PdfSharp/Drawing.Pdf/XGraphicsPdfRenderer.cs

Zaspard committed:
Refactored XImageBrush.cs and XGraphicsPdfRenderer:
  -XImageBrush:
    -Added info header to file
    -XImage is always created
  -XGraphicsPdfRenderer:
    -Added XImageBrush implementation when saving to PDF for WPF (tested) and GDI (not tested because NRE in BeginContainer)

Zaspard commented:
I fixed the WPF implementation and checked that it works when saving to a PDF file. With GDI, however, there are problems. 


In order to clip an area created by an image, I create a container (new XGraphicsContainer), make a SetClip on the passed object (XGraphicsPath), fill the area with images and close the container.  In the GDI implementation I need GraphicsState to create a container, but as far as I understand, it may not exist. What should I do in this case (file XGraphicsPdfRenderer.cs)?
```
                XGraphicsContainer container;
#if GDI
                container = new XGraphicsContainer(_gfx._gfx.Save());
#else
                container = new XGraphicsContainer();
#endif
                BeginContainer(container, xRect.Value, xRect.Value, XGraphicsUnit.Point);
                //I think it should be XCombineMode.Intersect
                SetClip(graphicsPath, XCombineMode.Intersect);
                for (int i = 0; i < xCount; i++)
                {
                    for (int j = 0; j < yCount; j++)
                    {
                        var xPosition = xRect.Value.X + (image.PixelWidth * i);
                        var yPosition = xRect.Value.Y + (image.PixelHeight * j);
                        DrawImage(image, xPosition, yPosition, image.PixelWidth, image.PixelHeight);
                    }
                }
                EndContainer(container);
```

Another question, as far as I understand in the GDI implementation you can't do this?
```
            // Create an empty page in this document.
            var page = document.AddPage();

            // Get an XGraphics object for drawing on this page.
            var gfx = XGraphics.FromPdfPage(page);

            var container = gfx.BeginContainer(); // <--throw exception in test
```
Message: 
  -System.InvalidOperationException : _gfx

Stack Trace: 
  -XGraphics.BeginContainer(XRect dstrect, XRect srcrect, XGraphicsUnit unit) line 4352
  -XGraphics.BeginContainer() line 4293
  -ImageTests.Create_brush_by_image() line 261

Zaspard committed:
Merge branch 'master' of https://github.com/Zaspard/PDFsharp into XImageBrush

Zaspard committed:
merge

Zaspard committed:
fixed merge + restored test + gdi still doesn't work

Zaspard committed:
Merge branch 'master' of https://github.com/Zaspard/PDFsharp into XImageBrush

Zaspard committed:
_gfx.BeginContainer for all implementation in DrawImageBrush

Zaspard commented:
Hello. I have not been able to solve the problem with creating XGraphicsContainer. Can you please tell me if there is any way to create XGraphicsContainer for GDI implementation? 

```
Message: 
-System.InvalidOperationException : _gfx

Stack Trace: 
-XGraphics.BeginContainer(XRect dstrect, XRect srcrect, XGraphicsUnit unit) line 4352
-XGraphics.BeginContainer() line 4293
-ImageTests.Create_brush_by_image() line 261
```

Best Regards,
Zaspard

Zaspard committed:
merge

Zaspard committed:
fixed merge

Zaspard commented:
Thanks for fixing the GDI bug

Best Regards,
Zaspard

Zaspard committed:
Merge branch 'master' of https://github.com/Zaspard/PDFsharp into XImageBrush

Zaspard committed:
Avoid incorrected area

Zaspard committed:
Refactor XImageBrush file
